### DISCUSSION TITLE
Schema implementation Ideas

### DISCUSSION THREAD

**Discussion by vicrdguez** [vicrdguez/dendroam#5] [Ideas]
This is an open discussion for implementing the [Dendron's concept of schemas](https://wiki.dendron.so/notes/c5e5adde-5459-409b-b34d-a0d75cbb1052.html).

Since there is not too much people interested on the project itself I'm going to use it as bucket for ideas and resources I found useful for this. However If anybody wants to contribute into this brain storming your ideas are more than welcome!

Right now what I have is:

# File type
## Use YAML
- Pros: Dendron compliant
- Cons: There is not too much support for YAML in emacs lisp

I've found these two libraries that might be useful:
- [emacs-libyaml](https://github.com/syohex/emacs-libyaml). Uses libyaml, looks like it's not being actively developed.
- [yaml.el](https://github.com/zkry/yaml.el) This is actively developed. Full elisp code, can be a bit slow

## JSON
Json is another option that is very similar to YAML (personally I like it more) and I think that could be easily exported to YAML if needed.
- Pros: native [elisp support](https://www.gnu.org/software/emacs/manual/html_node/elisp/Parsing-JSON.html)
- Cons: I don't find any right now

## Elisp objects
Another option can be just use elisp objects to represent schemas. This probably is the most convenient way in order to process values.
- Pros: Probably easier to work with in elisp
- Cons: difficult to work with for newcomers to emacs or people that don't know elisp

# Schema location
I think schemas should live in the same folder as notes, so the can be version controlled along with them and are portable. This is what Dendron does.

I think as well that schemas should be in separates files, having all of them in one file will be tedious with a big number of schemas.

# Schema UI
What I was thinking is to add this potential siblings of your note in the minibuffer as a candidate, with a marker indicating that the note doesn't exists but that the system is expecting it. Something like:
```
proj.project3.tasks         => this one exists and is part of the schema
proj.project3.notes [s]     => this one doesn't exists but is part of the schema ([s])
```
But this shouldn't stop you from creating a note outside of your schema if you need it, and the minibuffer can mark it as well
```
proj.project3.foo [?]      => Tells you that the note is not part of the schema
```

For this matter I think we need to use a dynamic completion list instead of an static one. For that reason I think `(completing-read)` won't work.
I guess the best bet will be something like [Ivy dynamic collections](http://oremacs.com/swiper/#example---counsel-locate), the way it is being used in `counsel-locate`.  This will allow to dynamically update the candidate list based on the search input, so we can show and hide potential childs that match the schema.

**Comment by nobiot** [comment]
@vicrdguez 
> For this matter I think we need to use a dynamic completion list instead of an static one. For that reason I think (completing-read) won't work.

Before going for what appears to be a complex solution with "Ivy dynamic collections", it might be still worth spending some time on `completing-read`.  I don't mean to steer your development in any particular way, and I have not really seen how Dendron works so I may have totally missed the mark here. 

But I believe that going with a simple and built-in framework with existing implementations might be worth considering before trying very hard to go with a custom solution. 

Let me illustrate what I mean below.

The POC I have done is based on `completing-read`, and yes the schema itself is fixed (but can be dynamically updated as it is just a variable that resides in memory). The visibility of the hierarchy changes dynamically, and accepts regex (such as "*", etc. -- in my case Orderless package) as you can see below.

![2021-04-27-dendroam-2](https://user-images.githubusercontent.com/12507865/116288304-ad07d400-a791-11eb-9492-b740c784f96d.gif)

In addition, the "marker" might also be possible with an existing facility of "annotation" in the standard Completion. Org-roam uses it and you can achieve something like this:

![](https://user-images.githubusercontent.com/12507865/112734570-e9ec6b00-8f46-11eb-9dc2-b58f82cb73bc.png)

The number of backlinks is dynamically generated by SQL query. In theory, I think you can check existence of files in the hierarchy and then display the marker. The tag is now part of the searchable text, so that part is obsolete (this is an old screen shot of mine).

These are built-in facilities, so you get them for free (not much custom scaffolding required).

Having said this, I completely understand if these do not represent your direction. Just wanted to give a perspective, as I really like what you are doing with dendroam 👍

**Reply by vicrdguez** [reply]
What you suggest here is very interesting @nobiot !
Very probably you are right with this and I'm thinking too complex.

Actually I'm learning elisp just now and I'm probably missing what can be done out of the box. That's exactly what I was looking for with this discussions!  Learning from other people and generate great ideas together.

I initially thought about dynamic collections when looking for ideas on how to avoid candidates clutter (showing more candidates than you really have) which I thought it could be a problem with a big amount of nodes of notes + headings. But reading your proposal maybe looks like it's not such a problem!

One of the things I wanted to add myself to this thread was your POC because was a great implementation itself but I haven't more time today, so thank you to bring it in. 

Talking about annotations, that's particularly what I was thinking of because I saw your post on that, and I thought it was just what was needed, so thank you again!

>Having said this, I completely understand if these do not represent your direction. Just wanted to give a perspective, as I really like what you are doing with dendroam 👍

Don't think about it, there is no really "my direction" I just have some guidelines of what I have in my head, but I made this public because I wanted to bring more people's opinions and build something useful not just for me.

Looking forward to keep reading your ideas!

**Comment by vicrdguez** [comment]
Adding @nobiot  POC for reference

https://gist.github.com/nobiot/152a5f0e6f761f5100e6b4e019bf0ccf

**Comment by vicrdguez** [comment]
I was further thinking how to approach this. I think the first step should be set 
how to define an schema, how will be the format, and how we are going to use them

This is a sample dendron schema:

``` yaml
version: 1
schemas:
# this will match "cli.*" notes
- id: cli 
  # human readable description of hierarchy
  desc: command line interface reference
  # add this to the domain of your schema hierarchy
  parent: root
  # when a schema is a namespace, it can have arbitrary children. equivalent to cli.* glob pattern
  namespace: true 
  children:
    - cmd
    - env
# will match cli.*.env
- id: env
  desc: cli specific env variables
# will match cli.*.cmd.*
- id: cmd
  desc: subcommands 
  namespace: true
```

So the main fields an schema have are:

- id (name of the hierarchy node)
- parent
- desc
- children
- namespace (indicates if the node can have arbitrary children or not)

I can identify this atomic unit description of a hierarchy element, the rest (children or parents) 
repeat the same structure.

If I would like to write this in elisp, it would be something like:

```emacs-lisp
'(:id "cli"
 :desc "command line interface reference"
 :parent "root"
 :namespace t
 :children ((:id "env"
             :desc "cli specific env variables")
            (:id "cmd"
             :desc "subcommands"
             :namespace: t)))
```

Now, assuming this is the schema, what it's left is, how to read them, and to 
inject potential candidates in the search list based on the input.

**Reply by vicrdguez** [reply]
I must say that it is easier to understand the schema in elisp than yaml IMHO haha
